x/gov/keeper/tally.go

```diff
@@ -10,23 +10,151 @@ import (
+	params, err := keeper.Params.Get(ctx)
+	if err != nil {
+		return false, false, v1.TallyResult{}, err
+	}
+	tallyResults = v1.NewTallyResultFromMap(results)
+
+	// If there is no staked coins, the proposal fails
+	totalBonded, err := keeper.sk.TotalBondedTokens(ctx)
+	if err != nil {
+		return false, false, v1.TallyResult{}, err
+	}
+
+	if totalBonded.IsZero() {
+		return false, false, tallyResults, nil
+	}
+
+	// If there are more spam votes than the sum of all other options, proposal fails
+	// A proposal with no votes should not be considered spam
+	if !totalVoterPower.Equal(math.LegacyZeroDec()) &&
+		results[v1.OptionSpam].GTE(results[v1.OptionOne].Add(results[v1.OptionTwo].Add(results[v1.OptionThree].Add(results[v1.OptionFour])))) {
+		return false, true, tallyResults, nil
+	}
+
+	switch proposal.ProposalType {
+	case v1.ProposalType_PROPOSAL_TYPE_OPTIMISTIC:
+		return keeper.tallyOptimistic(totalVoterPower, totalBonded, results, params)
+	case v1.ProposalType_PROPOSAL_TYPE_EXPEDITED:
+		return keeper.tallyExpedited(totalVoterPower, totalBonded, results, params)
+	case v1.ProposalType_PROPOSAL_TYPE_MULTIPLE_CHOICE:
+		return keeper.tallyMultipleChoice(totalVoterPower, totalBonded, results, params) // TODO(@julienrbrt): implement in follow up
+	default:
+		return keeper.tallyStandard(totalVoterPower, totalBonded, results, params)
+	}
+}
+
+func (keeper Keeper) tallyStandard(totalVoterPower math.LegacyDec, totalBonded math.Int, results map[v1.VoteOption]math.LegacyDec, params v1.Params) (passes, burnDeposits bool, tallyResults v1.TallyResult, err error) {
+	tallyResults = v1.NewTallyResultFromMap(results)
+
+	// If there is not enough quorum of votes, the proposal fails
+	percentVoting := totalVoterPower.Quo(math.LegacyNewDecFromInt(totalBonded))
+	quorum, _ := math.LegacyNewDecFromStr(params.Quorum)
+	if percentVoting.LT(quorum) {
+		return false, params.BurnVoteQuorum, tallyResults, nil
+	}
+
+	// If no one votes (everyone abstains), proposal fails
+	if totalVoterPower.Sub(results[v1.OptionAbstain]).Equal(math.LegacyZeroDec()) {
+		return false, false, tallyResults, nil
+	}
+
+	// If more than 1/3 of voters veto, proposal fails
+	vetoThreshold, _ := math.LegacyNewDecFromStr(params.VetoThreshold)
+	if results[v1.OptionNoWithVeto].Quo(totalVoterPower).GT(vetoThreshold) {
+		return false, params.BurnVoteVeto, tallyResults, nil
+	}
+
+	// If more than 1/2 of non-abstaining voters vote Yes, proposal passes
+	threshold, _ := math.LegacyNewDecFromStr(params.GetThreshold())
+
+	if results[v1.OptionYes].Quo(totalVoterPower.Sub(results[v1.OptionAbstain])).GT(threshold) {
+		return true, false, tallyResults, nil
+	}
+
+	// If more than 1/2 of non-abstaining voters vote No, proposal fails
+	return false, false, tallyResults, nil
+}
+
+func (keeper Keeper) tallyExpedited(totalVoterPower math.LegacyDec, totalBonded math.Int, results map[v1.VoteOption]math.LegacyDec, params v1.Params) (passes, burnDeposits bool, tallyResults v1.TallyResult, err error) {
```

Comment:
godoc

Reply:
could we get a bit more information on what happens in this function. They have different logic, we should leave a few sentences about what it does. It will be easier for the next person